File: MPI/answers/ex11/one_sided.f90
program basic
  implicit none
  include 'mpif.h'
  integer, parameter :: size = 100
  integer :: rc, myid, ntasks
  integer :: message(size)
  integer :: receiveBuffer(size)
  
  integer(kind=MPI_ADDRESS_KIND) :: winsize, disp
  integer :: window

  call mpi_init(rc)
  call mpi_comm_rank(MPI_COMM_WORLD, myid, rc)
  call mpi_comm_size(MPI_COMM_WORLD, ntasks, rc)

  message = myid

  winsize = sizeof(receiveBuffer) * size
  ! Create window corresponding to the receive buffer
  call mpi_win_create(receiveBuffer, winsize, sizeof(receiveBuffer), &
                      MPI_INFO_NULL, MPI_COMM_WORLD, window, rc)

  ! Send and receive as defined in the assignment
  call mpi_win_fence(0, window, rc)
  if ( myid < ntasks-1 ) then
     disp = 0
     call mpi_put(message, size, MPI_INTEGER, myid + 1, disp, size, &
                  MPI_INTEGER, window, rc)
     write(*,'(A,I0,A,I0,A,I0)') 'Origin: ', myid, &
          ' Put elements: ',size, &
          '. Target: ', myid+1
  end if
  call mpi_win_fence(0, window, rc)

  if ( myid > 0 ) then
     write(*,'(A,I0,A,I0)') 'Target: ', myid, &
          ' First element: ', receiveBuffer(1)
  end if

  call mpi_win_free(window, rc)
  call mpi_finalize(rc)

end program basic

File: MPI/answers/ex11/one_sided.c
#include<stdio.h>
#include<stdlib.h>
#include<mpi.h>


int main(int argc, char *argv[])
{
    int i, myid, ntasks;
    int size = 100;
    int *message;
    int *receiveBuffer;
    MPI_Win window;

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &ntasks);
    MPI_Comm_rank(MPI_COMM_WORLD, &myid);

    /* Allocate message */
    message = malloc(sizeof(int) * size);
    receiveBuffer = malloc(sizeof(int) * size);
    /* Initialize message */
    for (i = 0; i < size; i++) {
        message[i] = myid;
    }

    /* Create window corresponding to the receive buffer */
    MPI_Win_create(receiveBuffer, sizeof(int) * size, sizeof(int),
                   MPI_INFO_NULL, MPI_COMM_WORLD, &window);

    /* Put messages as defined in exercise */
    MPI_Win_fence(0, window);
    if (myid < ntasks - 1) {
        MPI_Put(message, size, MPI_INT, myid + 1, 0, size, MPI_INT,
                window);
        printf("Origin: %d. Put elements: %d. Target: %d\n", myid, size,
               myid + 1);
    }
    MPI_Win_fence(0, window);

    if (myid > 0) {
        printf("Target: %d. first element %d.\n", myid, receiveBuffer[0]);
    }

    free(message);
    free(receiveBuffer);
    MPI_Win_free(&window);
    MPI_Finalize();
    return 0;
}
